File: sim/golden_models/models/AfeAd711xxModel.cpp
#include "golden_models/models/AfeAd711xxModel.h"

#include <algorithm>

#include "golden_models/core/TestVectorIO.h"

namespace fpd::sim {

namespace {

uint32_t ComboDefaultNCols(uint32_t combo) {
    switch (combo) {
        case 4U:
        case 5U:
            return 1664U;
        case 6U:
        case 7U:
            return 3072U;
        default:
            return 2048U;
    }
}

uint32_t MinTLine(uint32_t afe_type) {
    return (afe_type == 1U) ? 6000U : 2200U;
}

uint32_t EffectiveIfs(uint32_t afe_type, uint32_t ifs) {
    return (afe_type == 1U) ? (ifs & 0x1FU) : (ifs & 0x3FU);
}

std::vector<uint16_t> MakeSampleLine(uint32_t channels, uint32_t seed) {
    std::vector<uint16_t> samples(channels);
    for (uint32_t index = 0; index < channels; ++index) {
        samples[index] = static_cast<uint16_t>((seed + index * 17U) & 0xFFFFU);
    }
    return samples;
}

}  // namespace

void AfeAd711xxModel::reset() {
    afe_start_ = 0;
    config_req_ = 0;
    cfg_combo_ = 1;
    cfg_tline_ = 0;
    cfg_ifs_ = 0;
    cfg_lpf_ = 0;
    cfg_pmode_ = 0;
    cfg_nchip_ = 1;
    afe_type_ = 0;
    afe_ready_ = 0;
    config_done_ = 0;
    dout_window_valid_ = 0;
    line_count_ = 0;
    tline_error_ = 0;
    ifs_width_error_ = 0;
    expected_ncols_ = 2048;
    sample_line_.clear();
    cycle_count_ = 0;
}

void AfeAd711xxModel::step() {
    config_done_ = 0;
    expected_ncols_ = ComboDefaultNCols(cfg_combo_);
    const uint32_t effective_tline = (cfg_tline_ == 0U) ? MinTLine(afe_type_) : cfg_tline_;
    const uint32_t effective_nchip = (cfg_nchip_ == 0U) ? 1U : cfg_nchip_;
    const uint32_t effective_ifs = EffectiveIfs(afe_type_, cfg_ifs_);
    tline_error_ = (effective_tline < MinTLine(afe_type_)) ? 1U : 0U;
    ifs_width_error_ = (afe_type_ == 1U && cfg_ifs_ > 31U) ? 1U : 0U;
    if (config_req_ != 0U && afe_ready_ == 0U) {
        afe_ready_ = 1U;
        config_done_ = 1U;
    }

    if (afe_start_ != 0U && afe_ready_ != 0U && tline_error_ == 0U) {
        dout_window_valid_ = 1U;
        ++line_count_;
        const uint32_t line_channels = std::min(expected_ncols_, 256U * effective_nchip);
        sample_line_ = MakeSampleLine(line_channels, effective_ifs + (effective_nchip << 8U));
        if (line_count_ >= effective_tline) {
            line_count_ = 0U;
            dout_window_valid_ = 0U;
        }
    } else {
        dout_window_valid_ = 0U;
        line_count_ = 0U;
    }
    ++cycle_count_;
}

void AfeAd711xxModel::set_inputs(const SignalMap& inputs) {
    afe_start_ = GetScalar(inputs, "afe_start", afe_start_);
    config_req_ = GetScalar(inputs, "config_req", config_req_);
    cfg_combo_ = GetScalar(inputs, "cfg_combo", cfg_combo_);
    cfg_tline_ = GetScalar(inputs, "cfg_tline", cfg_tline_);
    cfg_ifs_ = GetScalar(inputs, "cfg_ifs", cfg_ifs_);
    cfg_lpf_ = GetScalar(inputs, "cfg_lpf", cfg_lpf_);
    cfg_pmode_ = GetScalar(inputs, "cfg_pmode", cfg_pmode_);
    cfg_nchip_ = GetScalar(inputs, "cfg_nchip", cfg_nchip_);
    afe_type_ = GetScalar(inputs, "afe_type", afe_type_);
}

SignalMap AfeAd711xxModel::get_outputs() const {
    return {
        {"config_done", config_done_},
        {"afe_ready", afe_ready_},
        {"dout_window_valid", dout_window_valid_},
        {"tline_error", tline_error_},
        {"ifs_width_error", ifs_width_error_},
        {"expected_ncols", expected_ncols_},
        {"cfg_mix", (EffectiveIfs(afe_type_, cfg_ifs_) ^ cfg_lpf_ ^ cfg_pmode_ ^ cfg_nchip_) & 0xFFU},
        {"line_pixels", sample_line_},
    };
}

std::vector<Mismatch> AfeAd711xxModel::compare(const SignalMap& rtl_outputs) const {
    return CompareSignalMaps(cycle(), get_outputs(), rtl_outputs);
}

void AfeAd711xxModel::generate_vectors(const std::string& output_dir) {
    TestVectorFile vectors;
    vectors.module_name = "afe_ad711xx";
    vectors.spec_name = "SPEC-FPD-005";
    vectors.clock_name = "aclk";
    vectors.signal_inputs = {"afe_start", "config_req", "cfg_combo", "cfg_tline", "cfg_ifs", "cfg_nchip", "afe_type"};
    vectors.signal_outputs = {
        "config_done", "afe_ready", "dout_window_valid", "tline_error", "ifs_width_error", "expected_ncols"
    };

    reset();
    set_inputs({
        {"config_req", 1U},
        {"cfg_combo", 1U},
        {"cfg_tline", 2200U},
        {"cfg_ifs", 3U},
        {"cfg_nchip", 1U},
        {"afe_type", 0U},
    });
    step();
    vectors.vectors.push_back({cycle(), {
        {"config_req", 1U}, {"cfg_combo", 1U}, {"cfg_tline", 2200U}, {"cfg_ifs", 3U}, {"cfg_nchip", 1U}, {"afe_type", 0U}
    }, get_outputs()});

    set_inputs({
        {"afe_start", 1U},
        {"cfg_combo", 6U},
        {"cfg_tline", 6000U},
        {"cfg_ifs", 37U},
        {"cfg_nchip", 12U},
        {"afe_type", 1U},
    });
    step();
    vectors.vectors.push_back({cycle(), {
        {"afe_start", 1U}, {"cfg_combo", 6U}, {"cfg_tline", 6000U}, {"cfg_ifs", 37U}, {"cfg_nchip", 12U}, {"afe_type", 1U}
    }, get_outputs()});

    WriteHexVectors(vectors, output_dir + "/afe_ad711xx_vectors.hex");
    WriteBinaryVectors(vectors, output_dir + "/afe_ad711xx_vectors.bin");
}

}  // namespace fpd::sim

File: sim/viewer/src/FpdSimViewer/Models/AfeAd711xxModel.cs
using FpdSimViewer.Models.Core;

namespace FpdSimViewer.Models;

public sealed class AfeAd711xxModel : GoldenModelBase
{
    private uint _afeStart;
    private uint _configReq;
    private uint _cfgCombo = 1U;
    private uint _cfgTLine;
    private uint _cfgIfs;
    private uint _cfgLpf;
    private uint _cfgPMode;
    private uint _cfgNChip = 1U;
    private uint _afeType;
    private uint _afeReady;
    private uint _configDone;
    private uint _doutWindowValid;
    private uint _lineCount;
    private uint _tlineError;
    private uint _ifsWidthError;
    private uint _expectedNCols = 2048U;
    private ushort[] _sampleLine = [];

    public override void Reset()
    {
        _afeStart = 0U;
        _configReq = 0U;
        _cfgCombo = 1U;
        _cfgTLine = 0U;
        _cfgIfs = 0U;
        _cfgLpf = 0U;
        _cfgPMode = 0U;
        _cfgNChip = 1U;
        _afeType = 0U;
        _afeReady = 0U;
        _configDone = 0U;
        _doutWindowValid = 0U;
        _lineCount = 0U;
        _tlineError = 0U;
        _ifsWidthError = 0U;
        _expectedNCols = 2048U;
        _sampleLine = [];
        CycleCount = 0;
    }

    public override void Step()
    {
        _configDone = 0U;
        _expectedNCols = ComboDefaultNCols(_cfgCombo);
        var effectiveTLine = _cfgTLine == 0U ? MinTLine(_afeType) : _cfgTLine;
        var effectiveNChip = _cfgNChip == 0U ? 1U : _cfgNChip;
        var effectiveIfs = EffectiveIfs(_afeType, _cfgIfs);

        _tlineError = effectiveTLine < MinTLine(_afeType) ? 1U : 0U;
        _ifsWidthError = _afeType == 1U && _cfgIfs > 31U ? 1U : 0U;

        if (_configReq != 0U && _afeReady == 0U)
        {
            _afeReady = 1U;
            _configDone = 1U;
        }

        if (_afeStart != 0U && _afeReady != 0U && _tlineError == 0U)
        {
            _doutWindowValid = 1U;
            _lineCount++;
            var lineChannels = Math.Min(_expectedNCols, 256U * effectiveNChip);
            _sampleLine = MakeSampleLine(lineChannels, effectiveIfs + (effectiveNChip << 8));
            if (_lineCount >= effectiveTLine)
            {
                _lineCount = 0U;
                _doutWindowValid = 0U;
            }
        }
        else
        {
            _doutWindowValid = 0U;
            _lineCount = 0U;
        }

        CycleCount++;
    }

    public override void SetInputs(SignalMap inputs)
    {
        _afeStart = SignalHelpers.GetScalar(inputs, "afe_start", _afeStart);
        _configReq = SignalHelpers.GetScalar(inputs, "config_req", _configReq);
        _cfgCombo = SignalHelpers.GetScalar(inputs, "cfg_combo", _cfgCombo);
        _cfgTLine = SignalHelpers.GetScalar(inputs, "cfg_tline", _cfgTLine);
        _cfgIfs = SignalHelpers.GetScalar(inputs, "cfg_ifs", _cfgIfs);
        _cfgLpf = SignalHelpers.GetScalar(inputs, "cfg_lpf", _cfgLpf);
        _cfgPMode = SignalHelpers.GetScalar(inputs, "cfg_pmode", _cfgPMode);
        _cfgNChip = SignalHelpers.GetScalar(inputs, "cfg_nchip", _cfgNChip);
        _afeType = SignalHelpers.GetScalar(inputs, "afe_type", _afeType);
    }

    public override SignalMap GetOutputs()
    {
        return new SignalMap
        {
            ["afe_type"] = _afeType,
            ["afe_ready"] = _afeReady,
            ["config_done"] = _configDone,
            ["dout_window_valid"] = _doutWindowValid,
            ["line_count"] = _lineCount,
            ["tline_error"] = _tlineError,
            ["ifs_width_error"] = _ifsWidthError,
            ["expected_ncols"] = _expectedNCols,
            ["cfg_mix"] = (EffectiveIfs(_afeType, _cfgIfs) ^ _cfgLpf ^ _cfgPMode ^ _cfgNChip) & 0xFFU,
            ["sample_line"] = _sampleLine,
            ["line_pixels"] = _sampleLine,
        };
    }

    private static uint ComboDefaultNCols(uint combo)
    {
        return combo switch
        {
            4U or 5U => 1664U,
            6U or 7U => 3072U,
            _ => 2048U,
        };
    }

    private static uint MinTLine(uint afeType)
    {
        return afeType == 1U ? 6000U : 2200U;
    }

    private static uint EffectiveIfs(uint afeType, uint ifs)
    {
        return afeType == 1U ? (ifs & 0x1FU) : (ifs & 0x3FU);
    }

    private static ushort[] MakeSampleLine(uint channels, uint seed)
    {
        var samples = new ushort[channels];
        for (var index = 0; index < channels; index++)
        {
            samples[index] = (ushort)((seed + (index * 17U)) & 0xFFFFU);
        }

        return samples;
    }
}
